doc: Add withthegrid documentation (#716)
TheThingsIndustries/lorawan-stack-docs: doc/content/integrations/cloud-integrations/withthegrid/_index.md

function handle(args: Arguments, exec: Exec): IncomingRequestResponse | void {
  
  if (args.event.type === 'incomingRequest') {
    
    const request = args.event.webRequest;

    if (request.method === 'POST' && request.url.path === '/') {

      if (request.body === undefined || request.body.type !== 'json') {
        return {
          statusCode: 400,
          body: ['parameter_error', 'Body is not of type JSON'],
        };
      }

      const payload = JSON.parse(JSON.stringify(request.body.data));
      if (payload.join_accept !== undefined) {
        return {
          statusCode: 204,
          body: ['TTN is registering!'],
        };
      }

      // Set receive timestamp
      let generatedAt = new Date();
      generatedAt.setMilliseconds(0);

      // pass parsing of the report to the right report type parser
      exec.parseReport({
        reportTypeHashId,
        payload: JSON.stringify({
          generatedAt,
          payload: request.body.data,
        }),
      });
           
      return { statusCode: 204 };
    }

    return {
      statusCode: 404,
      body: ['not_found_error', 'Route cannot be found'],
    };
  }
}
```

> ## Step 7: Configuring TTN and registering the device

Reviewer: ```suggestion
## Step 7: Configuring {{% tts %}}
```

> Now that the webhook and the event handler of Dragino - LWL02 are configured, we can set up  TTN project. 

Reviewer: ```suggestion
Now that the webhook and the event handler of the Dragino - LWL02 device are configured, we can proceed with setting up {{% tts %}}. 
```

{{< figure src="Add_app.png" alt="TTN add app" >}}

> The second step is to add end device to the newly created application. For this you will need the AppEUI, DevEUI and the AppKey which is provided by the Dragino. After the device is registered, we can configure the application server by creating a new webhook. Set the webhook format to JSON and enter the Base URL listed in the Withthegrid webhook of TTN (step 2).

Reviewer: ```suggestion
[Create an application]({{< ref "/integrations/adding-applications" >}} and [add the Dragino device]({{< ref "/devices/adding-devices" >}}.

After the device is registered, we can configure a new webhook on {{% tts %}}. Set the webhook format to **JSON** and enter the **Base URL** provided with the Withthegrid webhook we created in step 2.
```

> {{< figure src="TTN_Webhook.png" alt="TTN set up webhook" >}}

Reviewer: This image should be larger, I think you can crop the bottom part with all these message types enabled. Btw, do you really want to enable all message types? Isn't this just for uplink messages?

## Step 8: Configuring report type

> The next step is to correctly configure the created report type such that it can decode the incoming payload into physical measurements. **The Things Network** makes this easy because the payload is already decoded. We only need to push the measurements to correct channels. Voltage measurement should be pushed into *Internal Sensor* channel which is the first channel, therefore, has a `channelIndex` of 0. Measurements related to *Probe* should be pushed to a `channelIndex` of 1.

Reviewer: ```suggestion
The next step is to correctly configure the created report type on Withthegrid, such that it can decode the incoming payload into physical measurements. {{% tts %}} makes this easy because the payload is already decoded. We only need to push the measurements to correct channels. Voltage measurement should be pushed into **Internal Sensor** channel which is the first channel, therefore, has a `channelIndex` of 0. Measurements related to **Probe** should be pushed to a `channelIndex` of 1.
```

Reviewer: You say that TTS makes this easy because the payload is decoded, but I don't see you're decoding data on TTS using payload formatters?

  // const lastLeakDuration = decodedPayload[6] << 16 | decodedPayload[7] << 8 | decodedPayload[8];
  
  const measurements: Result['measurements'] = [];

  pushMeasurement(measurements, sensorConditionChannelIndex, operatingVoltageQuantityHashId, generatedAt, -3, operatingVoltage * 1000);
  pushMeasurement(measurements, leakChannelIndex, leakStatusQuantityHashId, generatedAt, 0, waterLeakStatus);
  pushMeasurement(measurements, leakChannelIndex, totalLeakEventsQuantityHashId, generatedAt, 0, totalOpenLeakEvents);
  pushMeasurement(measurements, leakChannelIndex, lastLeakDurationQuantityHashId, generatedAt, 0, lastLeakDuration);
 
  return {
    generatedAt,
    measurements,
    fields: { },
  };
}

function pushMeasurement(measurements: Result['measurements'], channelIndex: number, quantityHashId: string, generatedAt: Date, orderOfMagnitude: number, significand: number) {
  measurements.push({
    channelIndex: channelIndex,
    quantityHashId: quantityHashId,
    generatedAt: generatedAt,
    orderOfMagnitude: orderOfMagnitude,
    significand: significand,
  });
}
```

## Step 9: Powering up the device

> After powering up the device it should connect to the application. You can see the successful connection in the connectivity environment in **Activity** and will look like this.

Reviewer: ```suggestion
After powering up the device it should automatically connect to the Withthegrid application. The successful connection can be observed in the connectivity environment under **Activity** and it will look like this:
```

{{< figure src="activity_1.png" alt="Activity" >}}

You can click on **Parse report** to view more information and verify that the parsing is correct.

{{< figure src="activity_2.png" alt="Snipet of parsed report" >}}

## Step 10: Claiming device tokens

> When the sensor is linked for the first time, the instance of the device is created in the platform. This can be found in **Devices**. The corresponding device will have a unique hashID (e.g. *dxrwmd*), Identifier (in our case IMEI of the device), and will be of device type: *Dragino - LWL02.* The device can be easily placed in any monitoring environment by selecting **GET CLAIM TOKEN(S)** in the bottom right corner.

Reviewer: ```suggestion
When the sensor is linked for the first time, the instance of the device is created in the platform and can be found in **Devices**. The corresponding device will have a unique hashID (e.g. **dxrwmd**), Identifier (in our case IMEI of the device), and will be of device type **Dragino - LWL02**. The device can be easily placed in any monitoring environment by selecting **GET CLAIM TOKEN(S)** in the bottom right corner.
```

{{< figure src="tokens_1.png" alt="Tokens for devices" >}}

Enter the hashID of a device to be claimed and click **REQUEST**.

{{< figure src="tokens_2.png" alt="HashID for claim token" >}}

The platform will output a claim token that can be copied.

{{< figure src="token_3.png" alt="Get claim token" >}}

In the last step, we will configure the monitoring environment and add the newly connected sensor. 

## Step 11: Configuring the monitoring environment

To add a device into the monitoring environment, navigate to **Devices**, click **+**, and enter the previously copied claim token. Click **ADD**.

{{< figure src="claim4.png" alt="Claim token" >}}

The following image shows that the device is registered in the monitoring environment, however, is not installed yet.

{{< figure src="Mon_1.png" alt="Claim device in monitoring environment" >}}

> To install a device, navigate to the created location (*ACME Headquarters*) and select Devices.

Reviewer: ```suggestion
To install a device, navigate to the created location (**ACME Headquarters**) and select **Devices**.
```

{{< figure src="claim7.png" alt="Create location" >}}

> In **Link to a device**, select the *Dragino - LWL02*. Finally, we would like to track the *Internal Sensor* port and *Probe* port onto which measurements are pushed. Click **+PORT** add Internal Sensor channel and repeat to add Probe channel. Finalize the device installation with **SET**.

Reviewer: ```suggestion
In **Link to a device**, select the **Dragino - LWL02**. Finally, we would like to track the **Internal Sensor**port and **Probe** port onto which measurements are pushed. Click **+ PORT** to add the **Internal Sensor** channel and repeat to add the **Probe** channel. Finalize the device installation with **SET**.
```

{{< figure src="Mon_2.png" alt="Set device ports" >}}

> From this point on, all measurements pushed to channels of *Dragino - LWL02* sensor with the unique IMEI will be available in the monitoring environments ready to be visualized in graphs. Try shorting the 2 pins on the sensor to see if a measurement is recorded.

Reviewer: ```suggestion
From this point on, all measurements pushed to channels of Dragino - LWL02 sensor with the unique IMEI will be available in the monitoring environments, ready to be visualized in graphs. Try shorting the two pins on the sensor to see if a measurement is recorded.
```
